plugins/kubernetes-backend/src/service/KubernetesProxy.ts

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright 2022 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { KubeConfig, bufferFromFileOrString } from '@kubernetes/client-node';
+import { Logger } from 'winston';
+import fetch from 'node-fetch';
+import * as https from 'https';
+
+import type { Request } from 'express';
+
+import {
+  ClusterDetails,
+  KubernetesProxyServices,
+  KubernetesClustersSupplier,
+} from '../types/types';
+
+const HEADER_CONTENT_TYPE: string = 'Content-Type';
+const APPLICATION_JSON: string = 'application/json';
+
+const HEADER_KUBERNETES_CLUSTERS: string = 'X-Kubernetes-Clusters';
+
+const ERROR_BAD_REQUEST: number = 400;
+const ERROR_NOT_FOUND: number = 404;
+const ERROR_INTERNAL_SERVER: number = 500;
+
+const CLUSTER_USER_NAME: string = 'backstage';
+
+export interface KubernetesProxyResponse {
+  code: number;
+  data: any;
+  cluster?: string;
+}
+
+interface KubernetesProxyClusters {
+  [key: string]: string;
+}
+
+export class KubernetesProxy {
+  constructor(protected readonly logger: Logger) {}
+
+  public async handleProxyRequest(
+    services: KubernetesProxyServices,
+    req: Request,
+  ): Promise<KubernetesProxyResponse> {
+    const krc = this.getKubernetesRequestedClusters(req);
+
+    if (Object.keys(krc).length < 1) {
+      return {
+        code: ERROR_NOT_FOUND,
```

Comment:
A hint here is that there are errors in `@backstage/errors` that backends use to signal problems to callers. Those are caught by the `errorHandler` middleware from backend-common if you add that to the route ([example](https://github.com/backstage/backstage/blob/8aa7d4f4098e6ec1ef922202bb95a08157d6a1ce/packages/backend-common/src/service/createStatusCheckRouter.ts#L53)). This is the standard way to deal with exceptional cases in backend plugins. Here you would do `throw new NotFoundError('No clusters found')`. This way the caller also gets rich error information, even including stack traces and stuff like that if it's been enabled. Just remember to add the middleware in the route builder above.

This would also be informative to the caller. You could separate out "hard" errors within the router code, from errors that are just propagated from the upstream.